doc/ERROR_CONCEPT.md

```diff
@@ -0,0 +1,250 @@
+* If I remember correctly: all error information is saved to this key, so users could investigate further error
+information which is saved to this key (in the metadata)
+
+Reasons against it:
+* Another extra line in the error message.
+* I highly doubt that most common users will require this information (unnecessary)
+
+Suggestions:
+Remove it for standard error messages. Maybe a `kdb set -v` could include such information for developers, users.
+Alternatively there could be a command for users to see the mountpoint, eg. `kdb get-mountpoint <key>`
+
+### Line 7: At
+
+Reasons to keep it:
+* Debug purposes
+
+Reasons against it:
+* Another extra line in the error message.
+* I highly doubt that most common users will require this information (unnecessary)
+
+Suggestions:
+Remove it for standard error messages. Maybe a `kdb set -v` could include such information for developers, users.
+Alternatively, another possibility would be to have a command which reveals the last error message with more detail: 
+`kdb last-message`.
+
+### Line 6: Module
+
+Reasons to keep it:
+* See the concrete plugin which causes the failure
+* Prohibits error masking (eg. `stat`ing a file happens in more plugins, devs will know where the error occurs)
+
+Reasons against it:
+* Another extra line in the error message.
+
+Suggestions:
+The need for plugin distinguishment in the error message should not be that high in my opinion. Maybe it would make sense to include it in line #1.
+
+### Line 5: Ingroup
+
+Reasons to keep it:
+* Another higher layer of distinguishment for the error location
+
+Reasons against it:
+* Extra line & unnecessary
+
+Suggestion:
+Having both ingroup and module is too much information. Also again in most cases users should get the relevant information out of the reason line.
+As of now there are 6 different "ingroup"s. I would suggest to remove it.
+
+### Line 4: Reason
+
+The most relevant line out of all, no question to keep it or not.
+Still I would remove the prefix `Reason: ` because of the reduced error message length in the new format. It makes it clear
+that this line is the reason. Probably we could substitute it with the plugin which called the error: 
+`<plugin>: Could not set ...`
+
+### Line 3: Description
+
+Reasons to keep it:
+* Gives users a quicker overview about the error (eg. permission error/ resource error/ etc. or another type of categorization such as Markus
+suggested with temporary/permanent/conflict/etc.)
+* Can indicate the type of solution a user has to approach
+
+Reasons against it:
+* Extra line
+* What to do with "uncategorizable" errors
+* Maintenance effort: introduce a new category once there are enough similar errors and change all old errors to this category?
+* Awareness of developers: All devs have to know all categories and choose the correct one. Devs should also be prohibited to throw their error
+into the "uncategorized" category due to lazyness.
+
```

Comment:
Maybe plugins should even say in the contract which errors they may throw?

Reply:
What is the advantage of having it in the contract?
Having a contract may seem to be useful but also causes even more overhead for developers and can cause code complication.
Personally for me it is not very intuitive to code in Elektra (specification, Metadata.ini, release_doc, internal functions, etc) and error contracts also takes it a little step further. But maybe this is common in the C world?

Reply:
It is similar to an exception specification: people know which errors they might get.

> also causes even more overhead for developers

I fully agree, so maybe better not have it.

> "and can cause code complication"

I do not understand.

> But maybe this is common in the C world?

No and I hope that your error concept will be easier to use for Elektra's developers.